jenkins/check_projects.py

```diff
@@ -0,0 +1,236 @@
+    ],
+}
+
+
+
+def Usage():
+  print >>sys.stderr, """Usage: %s [config] [--fix]
+  config: --default or json config path with {<ROLE>: [MEMBER, ...]}
+    {
+      "roles/editor": [
+        "serviceAccount:foo@bar.com",
+        "user:me@you.com",
+        "group:whatever@googlegroups.com"],
+      "roles/viewer": [...],
+      ...
+    }
+  --fix: try to add missing memberships if set
+  """ % (os.path.basename(sys.argv[0]))
+  sys.exit(1)
+
+
+def CheckProjects(configuration=None, fix=False):
+  if not configuration:
+    Usage()
+  if configuration == '--default':
+    needed = DEFAULT
+  elif not os.path.isfile(configuration):
+    Usage()
+  else:
+    with open(configuration) as fp:
+      needed = json.loads(fp.read())
+      bad_config = False
+      for role, members in DEFAULT.items():
+        memberships = set(needed.get(role, []))
+        missing = set(members) - set(needed.get(role, []))
+        if missing:
+          bad_config = True
+          print >>sys.stderr, 'Missing required memberships for %s:' % role
+          for member in missing:
+            print >>sys.stderr, '  %s' % member
+      if bad_config:
+        sys.exit(1)
+
+  projects = LoadProjects(os.path.dirname(__file__) or '.')
+  print >>sys.stderr, 'Checking %d projects for iam bindings:' % len(projects)
+  for role, needs in sorted(needed.items()):
+    print >>sys.stderr, '  %s: %s' % (role, ','.join(needs))
+
+  mutate = fix == '--fix'
+  threads = [
+      threading.Thread(target=Check, args=(p, needed, mutate))
+      for p in sorted(projects)
+  ]
+
+  for thread in threads:
+    thread.start()
+
+  for thread in threads:
+    thread.join()
+
+  print >>sys.stderr, '%d already configured, %d updated, %d problems' % (
+      len(completed), len(updated), len(errors))
+
+  fixers = collections.defaultdict(int)
+  unk = ['user:unknown']
+  for project in errors:
+    for name in helpers.get(project,unk):
+      fixers[name] += 1
+    print >>sys.stderr, '  %s: %s' % (
+        project, ','.join(Sane(s) for s in sorted(helpers.get(project, unk))))
+
+  print >>sys.stderr, 'Helpers:'
+  for name, count in sorted(fixers.items(), key=lambda i: i[1]):
+    print >>sys.stderr, '  %s: %s' % (count, Sane(name))
+  sys.exit(1)
+
+
+def Sane(member):
+  email = member.split(':')[1]
+  return email.split('@')[0]
+
+
+def LoadProjects(configs):
+  projects = set()
+  for dirname, _, files in os.walk(os.path.dirname(__file__) or '.'):
+    for path in files:
+      if not path.endswith('.yaml'):
+        continue
+      with open(os.path.join(dirname, path)) as fp:
+        for project in re.findall(r'PROJECT="(.+)"', fp.read()):
+          if '{' not in project:
+            projects.add(project)
+            continue
+          elif '{version-infix}' not in project:
+            with lock:
+              print >>sys.stderr, 'project expansion not allowed ',
+              print project
+              errors.add(project)
+          for version in [
```

Comment:
By not changing them when the next version comes out (we won't need to test some of the previous versions).

The real problem here is that we use 1-0, 1-1, 1-2 instead of something like deprecated, stable, current, next or thereabouts. Ideally we should (TM) fix that instead.